File: leetcode/206/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
     public ListNode reverseList(ListNode head) {
        if (head == null) return head;
        Stack<ListNode> stack = new Stack<ListNode>();
        ListNode cur = head;
        do {
            stack.push(cur);
            cur = cur.next;
        }
        while (cur != null);
        head = cur = stack.peek();
        while (!stack.isEmpty()) {
            cur.next = stack.pop();
            cur = cur.next;
        }
        cur.next = null;
        return head;
    }
}

File: leetcode/206/Solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head)
{
    if (head == NULL || head->next == NULL) return head;
    struct ListNode* prev = head;
    struct ListNode* cur = head->next;
    struct ListNode* post = cur->next;
    while (cur)
    {
        cur->next = prev;
        prev = cur;
        cur = post;
        if (post != NULL) post = cur->next; //队列尾部时post为空，需要特殊判断一下
    }
    head->next = NULL;//原来头指针再改一下即可
    return prev;
};